akka-typed/src/main/scala/akka/typed/internal/ActorCell.scala

```diff
@@ -0,0 +1,409 @@
+        system.eventStream.publish(Dropped(msg, self))
+      } else if (ActorCell.isTerminating(old)) {
+        if (Debug) println(s"[$thread] NOT enqueueing $msg into $self at status $old (is terminating)")
+        unsafe.getAndAddInt(this, statusOffset, -1)
+        system.deadLetters ! msg
+      } else {
+        if (Debug) println(s"[$thread] enqueueing $msg into $self at status $old")
+        // need to enqueue; if the actor sees the token but not the message, it will reschedule
+        queue.add(msg)
+        if (oldActivations == 0) {
+          if (Debug) println(s"[$thread] waking up $self")
+          unsafe.getAndAddInt(this, statusOffset, 1) // the first 1 was just the “active” bit, now add 1msg
+          // if the actor was not yet running, set it in motion; spurious wakeups don’t hurt
+          executionContext.execute(this)
+        }
+      }
+    } catch handleException
+
+  def sendSystem(signal: SystemMessage): Unit = {
+    @tailrec def needToActivate(): Boolean = {
+      val currentList = _systemQueue
+      if (currentList.head == NoMessage) {
+        system.deadLetters.toImpl.sendSystem(signal)
+        false
+      } else {
+        unsafe.compareAndSwapObject(this, systemQueueOffset, currentList.head, (signal :: currentList).head) || {
+          signal.unlink()
+          needToActivate()
+        }
+      }
+    }
+    try {
+      if (needToActivate()) {
+        val old = unsafe.getAndAddInt(this, statusOffset, 1)
+        if (isClosed(old)) {
+          // nothing to do
+          if (Debug) println(s"[$thread] NOT enqueueing $signal to $self: terminating")
+        } else if (activations(old) == 0) {
+          // all is good: we signaled the transition to active
+          if (Debug) println(s"[$thread] enqueueing $signal to $self: activating")
+          executionContext.execute(this)
+        } else {
+          // take back that token: we didn’t actually enqueue a normal message and the actor was already active
+          if (Debug) println(s"[$thread] enqueueing $signal to $self: already active")
+          unsafe.getAndAddInt(this, statusOffset, -1)
+        }
+      } else if (Debug) println(s"[$thread] NOT enqueueing $signal to $self: terminated")
+    } catch handleException
+  }
+
+  override final def run(): Unit = {
+    if (Debug) println(s"[$thread] entering run($self): interrupted=${Thread.interrupted()}")
+    val status = _status
+    val msgs = messageCount(status)
+    var processed = 0
+    try {
+      unscheduleReceiveTimeout()
+      if (!isClosed(status)) {
+        while (processAllSystemMessages() && !queue.isEmpty() && processed < msgs) {
+          val msg = queue.poll()
+          processed += 1
+          processMessage(msg)
+        }
+      }
+      scheduleReceiveTimeout()
+    } catch {
+      case NonFatal(ex) ⇒ fail(ex)
+      case ie: InterruptedException ⇒
+        fail(ie)
+        if (Debug) println(s"[$thread] interrupting due to catching InterruptedException")
```

Comment:
perhaps include `$self in all these debug messages with consistent format instead of sometimes ad-hoc
e.g. `s"[$thread] $self: interrupting due to catching InterruptedException"`